### PR TITLE
Add Task Cancellation

### PR DESCRIPTION
**Pull request by JosephEarl**
Initial CancellationToken & CancellationTokenSource implementation to match the C# API ([system.threading.cancellationtoken](https://msdn.microsoft.com/en-us/library/system.threading.cancellationtoken(v=vs.110).aspx) and [system.threading.cancellationtokensource](https://msdn.microsoft.com/en-us/library/system.threading.cancellationtokensource(v=vs.110).aspx)) and the cancellation API on Android (https://github.com/BoltsFramework/Bolts-Android/pull/53 and https://github.com/BoltsFramework/Bolts-Android/pull/56).

### CODE REVIEW

**Comment by nlutsenko-fb** [comment]
@JosephEarl This looks amazing!
Sorry for a swarm of comments, but all of these are nits.
Will happily merge this in once all of the nits are resolved.

Keep it up and i am looking really forward to see more of these awesome things!

**Comment by dimazen** [comment]
Is there any movements on this PR? I'd love to have cancellation support :sun_with_face:

**Comment by JosephEarl** [comment]
I'll be updating it shortly

**Comment by dimazen** [comment]
@JosephEarl thanks!

**Comment by JosephEarl** [comment]
Updated. Also added the ability to register a delegate to be notified when a token is cancelled.

**Comment by JosephEarl** [comment]
@nlutsenko-fb can you re-run the build please?

> No output has been received in the last ... minutes, this potentially indicates a stalled build or something wrong with the build itself.
> The build has been terminated

**Comment by nlutsenko-fb** [comment]
Doing this right now... Thanks!

**Comment by nlutsenko-fb** [comment]
@JosephEarl this looks even better!
In addition to nits (I am sorry about them) - I would recommend running a auto-indentation on every new file.

Also - it looks like there is a test failing :(

**Comment by JosephEarl** [comment]
* Added copyright
* Formatted new code
* Changed registerDelegate to registerCancellationObserverWithBlock
* Typedef'd BFCancellationBlock

File: Bolts/Common/BFCancellationTokenRegistration.h [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+ *  This source code is licensed under the BSD-style license found in the
+ *  LICENSE file in the root directory of this source tree. An additional grant
+ *  of patent rights can be found in the PATENTS file in the same directory.
+ *
+ */
+
+#import <Foundation/Foundation.h>
+
+/*!
+ Represents the registration of a delegate with a cancellation token.
```
Comment (JosephEarl) [inline Bolts/Common/BFCancellationTokenRegistration.h]: > registration of a delegate 

registration of a cancellation observer

File: Bolts/Common/BFCancellationTokenRegistration.h [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+#import <Foundation/Foundation.h>
+
+/*!
+ Represents the registration of a delegate with a cancellation token. 
+ Can be used to unregister the delegate at a later time.
+ */
+@interface BFCancellationTokenRegistration : NSObject
+
+/*!
+ Removes the block registered with the token and releases all resources associated
```
Comment (JosephEarl) [inline Bolts/Common/BFCancellationTokenRegistration.h]: > Removes the block

Removes the cancellation observer

**Comment by nlutsenko** [comment]
Hey @JosephEarl looks it won't merge automatically (needs rebasing).
Would be awesome to finally have this feature :yum:

**Comment by JosephEarl** [comment]
Will do

**Comment by JosephEarl** [comment]
Rebased and moved logic from CancellationTokenSource to CancellationToken, to match TaskCompletionSource and Task.

**Comment by nlutsenko** [comment]
@JosephEarl This looks great! Few comments, but I am very tempted to merge it in...

**Comment by JosephEarl** [comment]
Will fix up this evening

**Comment by JosephEarl** [comment]
Updated the docs and removed the token getter

**Comment by nlutsenko** [comment]
This looks great! Merging in!

**Comment by simonlinj** [comment]
Great job! thanks! 

I suggest the "Task Cancellation" described in the Readme.md file should be updated with this implementation.

**Comment by caoer** [comment]
great addition! :)

@nlutsenko @JosephEarl can I suggestion to make the cancellation token as the third parameter instead of 2nd? It can make the chaining method looks better. not sure how Android or C# doing with it.

change **FROM**
```
- (instancetype)continueWithExecutor:(BFExecutor *)executor
                               block:(BFContinuationBlock)block
                   cancellationToken:(BFCancellationToken *)cancellationToken;
```

**TO**
```
- (instancetype)continueWithExecutor:(BFExecutor *)executor
                   cancellationToken:(BFCancellationToken *)cancellationToken;
                               block:(BFContinuationBlock)block
```

**Comment by nlutsenko** [comment]
We can do an extra one, since these are already here and we probably don't want to break existing users :(
But a great suggestion, love the trailing block (also it maps waaay better with trailing closures in swift)

**Comment by JosephEarl** [comment]
Yes was just about to say we don't want to break the API. Sounds like a good addition.

**Pull request merged**